### DATA PT14-extra - Proyecto final - Grupo 3
#### Control de calidad de los datos del archivo Yelp reviews

In [1]:
import glob
import matplotlib.pyplot as plt
import numpy as np
import os
import pandas as pd
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

In [23]:
# Obtener la lista de archivos JSON en carpeta
#archivos_json = glob.glob(os.path.join("datos/yelp", "*.json"))

In [4]:
archivo = r'C:\Users\gonza\OneDrive\Desktop\Proyecto Grupal Henry\Yelp\review.json'
chunks = pd.read_json(archivo, lines=True, chunksize=100000)  # Leer en bloques de 100,000 filas

# Guardar cada chunk en un archivo pickle
# Esto evita que la memoria se llene al procesar el JSON por partes.
for i, chunk in enumerate(chunks):
    chunk.to_pickle(r'C:\Users\gonza\OneDrive\Desktop\Proyecto Grupal Henry\Yelp\review_part_{i}.pkl')  # Guarda en archivos más pequeños
    print(f'Chunk {i} guardado.')

Chunk 0 guardado.
Chunk 1 guardado.
Chunk 2 guardado.
Chunk 3 guardado.
Chunk 4 guardado.
Chunk 5 guardado.
Chunk 6 guardado.
Chunk 7 guardado.
Chunk 8 guardado.
Chunk 9 guardado.
Chunk 10 guardado.
Chunk 11 guardado.
Chunk 12 guardado.
Chunk 13 guardado.
Chunk 14 guardado.
Chunk 15 guardado.
Chunk 16 guardado.
Chunk 17 guardado.
Chunk 18 guardado.
Chunk 19 guardado.
Chunk 20 guardado.
Chunk 21 guardado.
Chunk 22 guardado.
Chunk 23 guardado.
Chunk 24 guardado.
Chunk 25 guardado.
Chunk 26 guardado.
Chunk 27 guardado.
Chunk 28 guardado.
Chunk 29 guardado.
Chunk 30 guardado.
Chunk 31 guardado.
Chunk 32 guardado.
Chunk 33 guardado.
Chunk 34 guardado.
Chunk 35 guardado.
Chunk 36 guardado.
Chunk 37 guardado.
Chunk 38 guardado.
Chunk 39 guardado.
Chunk 40 guardado.
Chunk 41 guardado.
Chunk 42 guardado.
Chunk 43 guardado.
Chunk 44 guardado.
Chunk 45 guardado.
Chunk 46 guardado.
Chunk 47 guardado.
Chunk 48 guardado.
Chunk 49 guardado.
Chunk 50 guardado.
Chunk 51 guardado.
Chunk 52 guardado.
Chu

In [5]:
import glob

# Obtener lista de archivos .pkl guardados
archivos_pkl = glob.glob(r'C:\Users\gonza\OneDrive\Desktop\Proyecto Grupal Henry\Yelp\review_part_*.pkl')

# Leer y concatenar los DataFrames
df_reviews = pd.concat([pd.read_pickle(f) for f in archivos_pkl])

print(df_reviews.head())

                      review_id                 user_id  \
6900000  BLu89xFjBFjcFB11ExqmbA  N3_CVCqI9bBVdrynd2KZVw   
6900001  0-JvMzZV1sy8mN5-b4mKng  nUbcXBO4tLHlm0SUHy9Qig   
6900002  yiBzu4GfyABEBImLR62uAw  _pNerxa8_oa2rGSsK8sckw   
6900003  wb3B-fwzK0JItr5spJMxLg  OIaACBkKofGy1QJw8hOYTg   
6900004  Qr_trHV78rMtTWJgTJhwEw  eWDVAz99_h2dSef6VF0Gzw   

                    business_id  stars  useful  funny  cool  \
6900000  jh8j-DWqgWkbRe_a2XtKFQ      5       0      0     0   
6900001  lhiaIa2-FkK1k5k-BeMxhg      5       3      0     0   
6900002  U1Gc3wKN5viu_e68KJVOdQ      5       3      1     2   
6900003  _OxsSJLOrhdJq3-EewouvQ      5       0      0     0   
6900004  Bg7D8LrsW9XbYlEhT9yekw      1       0      0     0   

                                                      text                date  
6900000  Terrific place, I've known some of the staff f... 2020-12-17 02:02:53  
6900001  My previous dentist died from a fast moving ca... 2018-07-06 01:14:47  
6900002  When it's 10PM

In [6]:
print('text NaN', df_reviews['text'].isna().sum())
print('stars NaN', df_reviews['stars'].isna().sum())

text NaN 0
stars NaN 0


El dataframe reviews de Yelp no tiene valores nulos para la columna text (con los comentarios de las reseñas) ni para la columna stars (con la calificacion) 

In [7]:
# Convertir la columna a tipo datetime
# Vamos a asegurarnos de que la columna "date" esté en el formato correcto:
df_reviews['date'] = pd.to_datetime(df_reviews['date'], errors='coerce')

In [18]:
df_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90280 entries, 6900000 to 6990279
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   review_id    90280 non-null  object        
 1   user_id      90280 non-null  object        
 2   business_id  90280 non-null  object        
 3   stars        90280 non-null  int64         
 4   useful       90280 non-null  int64         
 5   funny        90280 non-null  int64         
 6   cool         90280 non-null  int64         
 7   text         90280 non-null  object        
 8   date         90280 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(4)
memory usage: 6.2+ MB


In [12]:
df_reviews.head()

review_id                 user_id  \
6900000  BLu89xFjBFjcFB11ExqmbA  N3_CVCqI9bBVdrynd2KZVw   
6900001  0-JvMzZV1sy8mN5-b4mKng  nUbcXBO4tLHlm0SUHy9Qig   
6900002  yiBzu4GfyABEBImLR62uAw  _pNerxa8_oa2rGSsK8sckw   
6900003  wb3B-fwzK0JItr5spJMxLg  OIaACBkKofGy1QJw8hOYTg   
6900004  Qr_trHV78rMtTWJgTJhwEw  eWDVAz99_h2dSef6VF0Gzw   

                    business_id  stars  useful  funny  cool  \
6900000  jh8j-DWqgWkbRe_a2XtKFQ      5       0      0     0   
6900001  lhiaIa2-FkK1k5k-BeMxhg      5       3      0     0   
6900002  U1Gc3wKN5viu_e68KJVOdQ      5       3      1     2   
6900003  _OxsSJLOrhdJq3-EewouvQ      5       0      0     0   
6900004  Bg7D8LrsW9XbYlEhT9yekw      1       0      0     0   

                                                      text                date  
6900000  Terrific place, I've known some of the staff f... 2020-12-17 02:02:53  
6900001  My previous dentist died from a fast moving ca... 2018-07-06 01:14:47  
6900002  When it's 10PM on a Friday during COVID times,... 2021-01-09 17:03:54  
6900003  Dane was outstanding and worked very hard to m... 2021-08-10 23:28:42  
6900004  Very small was looking for the Van Gogh exhibi... 2021-08-11 00:49:25

El dataframe tiene 6.990.280 reviews. La columna date esta en formato datetime.

In [19]:
# df_rev.to_parquet('/yelp/reviews.parquet' , engine='fastparquet')
# El motor pyarrow maneja mejor archivos grandes y evita problemas de compresión:
df_reviews.to_parquet(r'C:\Users\gonza\OneDrive\Desktop\Proyecto Grupal Henry\Yelp\yelp_reviews.parquet', engine='pyarrow', compression='snappy')

In [20]:
df_dup = df_reviews[df_reviews.duplicated(subset=['business_id',  'date', 'review_id', 'user_id'], keep=False)]
df_dup = df_dup.sort_values([])

In [21]:
print(df_dup)

Empty DataFrame
Columns: [review_id, user_id, business_id, stars, useful, funny, cool, text, date]
Index: []


No hay filas duplicadas

In [22]:
df_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90280 entries, 6900000 to 6990279
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   review_id    90280 non-null  object        
 1   user_id      90280 non-null  object        
 2   business_id  90280 non-null  object        
 3   stars        90280 non-null  int64         
 4   useful       90280 non-null  int64         
 5   funny        90280 non-null  int64         
 6   cool         90280 non-null  int64         
 7   text         90280 non-null  object        
 8   date         90280 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(4)
memory usage: 6.2+ MB


In [25]:
# leo el archivo de las pizzerias mt_pizza_NJNY para hacer join con este
# How to read a Parquet file into Pandas DataFrame?
business_pizza_CANV = pd.read_parquet(r'C:\Users\gonza\OneDrive\Desktop\Proyecto Grupal Henry\Maps\df_pizza_CANV_rev.parquet' , engine='fastparquet')

In [26]:
business_pizza_CANV.info()

<class 'pandas.core.frame.DataFrame'>
Index: 34905 entries, 0 to 36420
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   gmap_id         34905 non-null  object        
 1   business_name   34905 non-null  object        
 2   address         34891 non-null  object        
 3   city            34905 non-null  object        
 4   state           34905 non-null  object        
 5   zip_code        34905 non-null  object        
 6   latitude        34905 non-null  float64       
 7   longitude       34905 non-null  float64       
 8   avg_rating      34905 non-null  float64       
 9   num_of_reviews  34905 non-null  int64         
 10  user_id         34905 non-null  float64       
 11  user_name       34905 non-null  object        
 12  rating          34905 non-null  float64       
 13  text            21393 non-null  object        
 14  date            34905 non-null  datetime64[ns]
dtypes: date

In [30]:
print(business_pizza_CANV.columns)
print(df_reviews.columns)


Index(['gmap_id', 'business_name', 'address', 'city', 'state', 'zip_code',
       'latitude', 'longitude', 'avg_rating', 'num_of_reviews', 'user_id',
       'user_name', 'rating', 'text', 'date'],
      dtype='object')
Index(['review_id', 'user_id', 'business_id', 'stars', 'useful', 'funny',
       'cool', 'text', 'date'],
      dtype='object')


In [32]:
business_pizza_CANV = pd.merge(business_pizza_CANV, df_reviews, how='left', left_on='business_name', right_on='business_id') # tengo que hacer left join



In [33]:
business_pizza_CANV.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34905 entries, 0 to 34904
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   gmap_id         34905 non-null  object        
 1   business_name   34905 non-null  object        
 2   address         34891 non-null  object        
 3   city            34905 non-null  object        
 4   state           34905 non-null  object        
 5   zip_code        34905 non-null  object        
 6   latitude        34905 non-null  float64       
 7   longitude       34905 non-null  float64       
 8   avg_rating      34905 non-null  float64       
 9   num_of_reviews  34905 non-null  int64         
 10  user_id_x       34905 non-null  float64       
 11  user_name       34905 non-null  object        
 12  rating          34905 non-null  float64       
 13  text_x          21393 non-null  object        
 14  date_x          34905 non-null  datetime64[ns]
 15  re

Hay 34905 reseñas de Yelp para las pizzerias del estado

In [34]:
# Modifico los nombres de las columnas que estan duplicados
business_pizza_CANV.rename(columns={'stars_x': 'stars_business', 'stars_y': 'stars_by_user'}, inplace=True)

In [35]:
business_pizza_CANV.head()

gmap_id      business_name       address  \
0  0x80c2c7726b53b8e5:0xad288d11654f5e48  PIZZA ROW RECORDS  433 E 7th St   
1  0x80c2c7726b53b8e5:0xad288d11654f5e48  PIZZA ROW RECORDS  433 E 7th St   
2  0x80c2c7726b53b8e5:0xad288d11654f5e48  PIZZA ROW RECORDS  433 E 7th St   
3  0x80c2c7726b53b8e5:0xad288d11654f5e48  PIZZA ROW RECORDS  433 E 7th St   
4  0x80c2c7726b53b8e5:0xad288d11654f5e48  PIZZA ROW RECORDS  433 E 7th St   

          city state zip_code   latitude   longitude  avg_rating  \
0  Los Angeles    CA    90014  34.041406 -118.247477         4.3   
1  Los Angeles    CA    90014  34.041406 -118.247477         4.3   
2  Los Angeles    CA    90014  34.041406 -118.247477         4.3   
3  Los Angeles    CA    90014  34.041406 -118.247477         4.3   
4  Los Angeles    CA    90014  34.041406 -118.247477         4.3   

   num_of_reviews  ...                  date_x review_id  user_id_y  \
0              34  ... 2019-08-10 16:42:31.213       NaN        NaN   
1              34  ... 2021-08-22 21:24:59.592       NaN        NaN   
2              34  ... 2019-09-13 07:05:09.213       NaN        NaN   
3              34  ... 2019-11-28 07:44:09.374       NaN        NaN   
4              34  ... 2019-08-25 14:23:22.205       NaN        NaN   

  business_id stars useful funny cool  text_y  date_y  
0         NaN   NaN    NaN   NaN  NaN     NaN     NaT  
1         NaN   NaN    NaN   NaN  NaN     NaN     NaT  
2         NaN   NaN    NaN   NaN  NaN     NaN     NaT  
3         NaN   NaN    NaN   NaN  NaN     NaN     NaT  
4         NaN   NaN    NaN   NaN  NaN     NaN     NaT  

[5 rows x 24 columns]

In [36]:
print(business_pizza_CANV.iloc[[0]])

                                 gmap_id      business_name       address  \
0  0x80c2c7726b53b8e5:0xad288d11654f5e48  PIZZA ROW RECORDS  433 E 7th St   

          city state zip_code   latitude   longitude  avg_rating  \
0  Los Angeles    CA    90014  34.041406 -118.247477         4.3   

   num_of_reviews  ...                  date_x review_id  user_id_y  \
0              34  ... 2019-08-10 16:42:31.213       NaN        NaN   

  business_id stars useful funny cool  text_y  date_y  
0         NaN   NaN    NaN   NaN  NaN     NaN     NaT  

[1 rows x 24 columns]


In [37]:
# guardo df_pizza_CANV_rev
business_pizza_CANV.to_parquet(r'C:\Users\gonza\OneDrive\Desktop\Proyecto Grupal Henry\Yelp\business_pizza_CANV_rev.parquet' , engine='fastparquet')